### ISSUE TITLE
suggested code to enable 'go vet' to report printf mistakes in the source code

### ISSUE THREAD

**Issue by adonovan** [kubernetes/kubernetes#114568]
### What happened?

(This isn't a bug report, but I couldn't find a more appropriate category.)

The k8s Go source code contains a number of places where printf-like functions are called with arguments of the wrong type. Ordinarily, the 'go vet' tool would point out such mistakes, but unfortunately it's not smart enough to figure out that loggingT.printfDepth behaves like printf, because it obscures the call delegation with extra computation. However, the trivial patch below would make the connection obvious:

```
 func (l *loggingT) printfDepth(s severity.Severity, logger *logr.Logger, filter LogFilter, depth int, format string, args ...interface{}) {
+       if false {
+               _ = fmt.Sprintf(format, args...) // cause vet to treat this function like printf
+       }
```

and cause vet to discover a number of mistakes, such as these:
```
kubernetes$ go vet ./...
# k8s.io/kubernetes/cmd/kubectl
cmd/kubectl/kubectl.go:35:2: k8s.io/kubectl/pkg/cmd/util.APrintf format %d has arg "123" of wrong type string
cmd/kubectl/kubectl.go:36:2: k8s.io/kubectl/pkg/cmd/util.Printf format %d has arg "123" of wrong type string
# k8s.io/kubernetes/cmd/kubelet
cmd/kubelet/kubelet.go:57:2: (*github.com/spf13/cobra.Command).PrintErrf format %d has arg "123" of wrong type string
cmd/kubelet/kubelet.go:66:2: k8s.io/klog/v2.Warningf format %d has arg "123" of wrong type string
cmd/kubelet/kubelet.go:67:6: fmt.Sprintf format %d has arg "123" of wrong type string
# k8s.io/kubernetes/pkg/controller/daemon
pkg/controller/daemon/update.go:163:5: (k8s.io/klog/v2.Verbose).Infof call needs 2 args but has 4 args
# k8s.io/kubernetes/pkg/controller/endpointslice/topologycache
pkg/controller/endpointslice/topologycache/topologycache.go:215:3: (k8s.io/klog/v2.Verbose).Infof format %s has arg totalCPU.MilliValue() of wrong type int64
# k8s.io/kubernetes/pkg/controller/job
pkg/controller/job/job_controller.go:939:19: (k8s.io/klog/v2.Verbose).Infof call has arguments but no formatting directives
# k8s.io/kubernetes/pkg/controller/statefulset
pkg/controller/statefulset/stateful_pod_control.go:239:4: (k8s.io/klog/v2.Verbose).Infof format %s reads arg #1, but call has 0 args
# k8s.io/kubernetes/pkg/controller/volume/persistentvolume
pkg/controller/volume/persistentvolume/pv_controller.go:803:6: k8s.io/klog/v2.Warningf call needs 1 arg but has 2 args
# k8s.io/kubernetes/pkg/volume/cinder
pkg/volume/cinder/cinder.go:476:3: k8s.io/klog/v2.Warningf format %w has arg dir of wrong type string
```

### What did you expect to happen?

.

### How can we reproduce it (as minimally and precisely as possible)?

.

### Kubernetes version

.

### Cloud provider

.

### OS version

.

**Comment by BenTheElder** [comment]
/sig instrumentation

I think this should be in the klog linter, we do NOT want to be forking go tools where avoidable. 

We're also in the direction of using golangci-lint over `go vet` directly.

**Comment by adonovan** [comment]
To be clear: I am suggesting you make an innocuous code change that would allow the unmodified `go vet` tool (or `go test`, which runs `go vet` in some cases), or any other Go analysis driver that embeds the `printf` checker, to detect mistakes in your Go corpus. I am not suggesting that you fork any code.

**Comment by BenTheElder** [comment]
🤦‍♂️, this is what I get for poking through issues before the caffeine hits

somewhat related:
/sig testing
/sig architecture
/area code-organization

That seems like a bit of an odd hack (which, admittedly the project is full of), I'd be somewhat surprised if a different linter didn't complain about unreachable code :+)

**Comment by adonovan** [comment]
An explicit "if false { ... }" block is a pretty traditional way to comment out debugging code while still subjecting it to type checking, cross-refs, etc, so I'd push back on any linter that told me I couldn't use that.

**Comment by pohly** [comment]
/assign

That looks useful.

As a side note: several of the findings are in code that is currently getting converted to structured, contextual logging. It would be good to get those changes merged first, then there will be less to clean up when enabling the check.

**Comment by BenTheElder** [comment]
@logicalhan is logging really not SIG Instrumentation?

**Comment by logicalhan** [comment]
> @logicalhan is logging really not SIG Instrumentation?

Isn't this a linting issue?

**Comment by BenTheElder** [comment]
I mean, it's suggesting an approach to catch misuse of logging printf type methods. Similar to https://github.com/kubernetes/kubernetes/issues/102439

**Comment by pohly** [comment]
I doesn't matter, does it? I can handle this for SIG Testing or for SIG Instrumentation... :grin:

**Comment by BenTheElder** [comment]
The source changes will require instrumentation approval, so I would
suggest instrumentation should also be tracking this.

Might be some misunderstanding, this issue doesn't require a new linter or
any linter changes, this requires changes at the logging method source.

**Comment by pohly** [comment]
I don't know whether it matters which SIG tracks this, but I agree, the code change falls under instrumentation, so:

/sig instrumentation

**Comment by dims** [comment]
@pohly do we have this already in CI jobs running in latest master?

**Comment by pohly** [comment]
No, not yet. But quite a bit of the logging calls have been cleaned up, so let me give it a try tomorrow.

**Comment by pohly** [comment]
We can and should do the same for `fmt.Println` and `fmt.Print` style functions (`klog.Infoln`, `klog.Info`):
```
k8s.io/klog/v2.Info call has possible formatting directive %s
```

Yes, klog has too many API calls...

klog change is in https://github.com/kubernetes/klog/pull/390, please review.

**Comment by pohly** [comment]
Resolved in https://github.com/kubernetes/klog/pull/390 and enabled in Kubernetes 1.29 via https://github.com/kubernetes/kubernetes/pull/121552.

/close
